docs/source/developer/extension_dev.rst

```diff
@@ -3,42 +3,204 @@
+  out of memory rebuilding JupyterLab, particularly in resource-limited
+  containers).
+- (new in JupyterLab 3.0, recommended) A "federated" JavaScript bundle. In
+  this case, the extension developer uses tools provided by JupyterLab to
+  compile the extension into a JavaScript bundle that includes the
+  non-JupyterLab JavaScript dependencies, then distributes the resulting
+  bundle in, say, a Python pip or conda package. JupyterLab can load these
+  federated extensions directly from the bundle without having to rebuild
+  itself. This means the end user does not have to have Node.js and can
+  immediately use the extension without a costly rebuild, but may use a little
+  more network bandwidth to load the federated extensions. 
+  
+An extension may be distributed as both a single JavaScript package published
+on NPM and as a federated extension bundle published in a Python package,
+giving users the choice of how to install it. Also, a user may install some
+extensions as single JavaScript packages (rebuilding JupyterLab to incoporate
+those in the base application) and some other extensions as federated bundles
+which are loaded dynamically in the browser.
+
+Because federated extensions do not require a JupyterLab rebuild, they have a
+distinct advantage in multiuser systems where JuptyerLab is installed at the
+system level. On such systems, only the system administrator has permissions
+to rebuild JupyterLab and install single JavaScript package extensions. Since
+federated extensions can be installed at the system level, per-environment
+level, or per-user level, each user can have their own separate set of
+federated extensions that are loaded dynamically in their browser on top of
+the system-wide JuptyerLab.
+
+.. tip::
+   We recommend extensions be distributed as federated bundles in Python
+   packages for user convenience.
+
+
+Writing an Extension
+--------------------
 
--  **application plugins (top level):** Application plugins extend the
-   functionality of JupyterLab itself.
--  **mime renderer extensions (top level):** Mime Renderer extensions are
-   a convenience for creating an extension that can render mime data and
-   potentially render files of a given type.
--  **theme extensions (top level):** Theme extensions allow you to customize the appearance of
-   JupyterLab by adding your own fonts, CSS rules, and graphics to the application.
--  **document widget extensions (lower level):** Document widget extensions
-   extend the functionality of document widgets added to the
-   application, and we cover them in :ref:`documents`.
 
-A JupyterLab application is comprised of:
+We encourage extension authors to add the `jupyterlab-extension GitHub topic
+<https://github.com/search?utf8=%E2%9C%93&q=topic%3Ajupyterlab-extension&type=Repositories>`__
+to any GitHub extension repository.
+
+
+Resources
+^^^^^^^^^
+
+Tutorials, cookie cutters, and examples
+
+
+
+package.json metadata
```

Comment:
Nice, this will be a nice place to document the JupyterLab Package Metadata schema :+1:  https://github.com/jupyterlab/jupyterlab/blob/master/builder/metadata_schema.json